linkerd/app/gateway/src/discover.rs

```diff
@@ -0,0 +1,111 @@
+                failfast_timeout: queue.failfast_timeout,
+            }
+        };
+        svc::mk(move |GatewayAddr(addr)| {
+            tracing::debug!(%addr, "Discover");
+
+            if !allowlist.matches(addr.name()) {
+                tracing::debug!(%addr, "Address not in gateway discovery allowlist");
+                return future::Either::Left(future::err(GatewayDomainInvalid.into()));
+            }
+
+            let profile = profiles
+                .clone()
+                .get_profile(profiles::LookupAddr(addr.clone().into()))
+                .instrument(tracing::debug_span!("profiles"));
+
+            let policy = policies
+                .get_policy(addr.clone().into())
+                .instrument(tracing::debug_span!("policy"));
+
+            let discovery = future::join(profile, policy).map(move |(profile, policy)| {
+                tracing::debug!("Discovered");
+
+                let profile = profile?.ok_or(GatewayDomainInvalid)?;
+
+                // If there was a policy resolution, return it with the profile so
+                // the stack can determine how to switch on them.
+                match policy {
+                    Ok(policy) => Ok((Some(profile), policy)),
+                    // The policy controller currently rejects discovery for pod
+                    // names (rather than service names) so we'll need to
+                    // synthesize a policy.
+                    Err(error) if is_not_found(&error) => {
+                        tracing::debug!("Policy not found");
+                        let policy = outbound::discover::spawn_synthesized_profile_policy(
+                            profile.clone().into(),
+                            move |profile| {
+                                static META: Lazy<Arc<policy::Meta>> = Lazy::new(|| {
+                                    Arc::new(policy::Meta::Default {
+                                        name: "gateway".into(),
+                                    })
+                                });
+
+                                match profile.endpoint.clone() {
+                                    Some((addr, meta)) => ClientPolicy::synthesize_forward(
+                                        &META,
+                                        detect_timeout,
+                                        queue,
+                                        addr,
+                                        meta,
+                                    ),
+                                    None => {
+                                        // TODO(eliza): should we synthesize a policy from the nameaddr?
+                                        tracing::debug!(
+                                            "Gateway ServiceProfile does not contain an endpoint"
+                                        );
+                                        ClientPolicy::invalid(detect_timeout)
+                                    }
```

Comment:
If the policy controller did not return a policy for the discovery target, it was probably a pod name, and therefore *should* be a forward...so it's *probably* fine to synthesize an invalid policy in the case where there's no forward endpoint, I think? Otherwise, I'm not really sure what we could do here that would be coherent...